crates/uv/src/commands/project/run.rs

```diff
@@ -882,6 +883,20 @@ impl TryFrom<&ExternalCommand> for RunCommand {
         {
             Ok(Self::PythonGuiScript(target_path, args.to_vec()))
         } else {
+            // Sniff the shebang to avoid infinite recursion
+            let known_shebang = br"#!/usr/bin/env -S uv run";
```

Comment:
I also worry this heuristic is too brittle. I'm not sure what the solution is without peeking. Some sort of environment variable?

Reply:
We also have something like... #3097 — maybe we just need a `--script` flag here to force us to treat something as a script? We should still have a way to detect recursion and bail instead of looping though.

Reply:
(i.e. a `UV_INTERNAL__RECURSION` variable that we bump on every recursive subprocess and bail after some limit is hit)

Reply:
A `--script` flag plus a `UV_INTERNAL_RECURSION` for better error reporting would work.

Reply:
A more resilient solution would be not running binaries, but only running python scripts and `python` from `uv run`, this seems to be done by e.g. `bash` to avoid infinite recursion.

Reply:
I'm pretty unwilling to change that part of the user experience — running binaries from `uv run` is important. I think a separate command or opt-in to that experience via `--script` makes the most sense to me.